src/many-ledger/src/migration/tokens.rs

```diff
@@ -0,0 +1,162 @@
+use crate::error;
+use crate::migration::MIGRATIONS;
+use crate::storage::ledger_tokens::{key_for_ext_info, key_for_symbol};
+use crate::storage::{SYMBOLS, SYMBOLS_BYTES};
+use linkme::distributed_slice;
+use many_error::ManyError;
+use many_identity::Address;
+use many_migration::InnerMigration;
+use many_modules::ledger::extended_info::TokenExtendedInfo;
+use many_types::ledger::{Symbol, TokenAmount, TokenInfo, TokenInfoSummary, TokenInfoSupply};
+use merk::Op;
+use serde_json::Value;
+use std::collections::{BTreeMap, HashMap};
+use std::str::FromStr;
+
+/// Migrate the subresource counter from "/config/account_id" to "/config/subresource_id"
+fn migrate_subresource_counter(storage: &mut merk::Merk) -> Result<(), ManyError> {
+    // Is the old counter present in the DB?
+    let old_counter = storage
+        .get(b"/config/account_id")
+        .map_err(|_| ManyError::unknown("Unable to retrieve old counter"))?;
+
+    // Is the new counter present in the DB?
+    let new_counter = storage
+        .get(b"/config/subresource_id")
+        .map_err(|_| ManyError::unknown("Unable to retrieve new counter"))?;
+
+    match (old_counter, new_counter) {
+        // Old counter is present, new counter is not. First time running the migration.
+        (Some(x), None) => {
+            // Migrate the old counter to the new location in the database
+            storage
+                .apply(&[(b"/config/subresource_id".to_vec(), Op::Put(x))])
+                .map_err(error::storage_apply_failed)?;
+
+            // Delete the old counter from the DB
+            storage
+                .apply(&[(b"/config/account_id".to_vec(), Op::Delete)])
+                .map_err(error::storage_apply_failed)?;
+        }
+        // No counter found. Set the new counter to 0.
+        (None, None) => {
+            storage
+                .apply(&[(b"/config/subresource_id".to_vec(), Op::Put(vec![0u8; 4]))])
+                .map_err(error::storage_apply_failed)?;
+        }
+        // Old counter is not present, new counter is present.
+        // The migration did run in the past.
+        // Skip this step
+        (None, Some(_)) => {}
+        // Both counters are present in the DB.
+        // Something wrong is happening
+        (Some(_), Some(_)) => {
+            return Err(ManyError::unknown(
+                "Two subresource counters found in the store; aborting",
+            ))
+        }
+    }
+    Ok(())
+}
+
+fn migrate_token_metadata(
+    storage: &mut merk::Merk,
+    extra: &HashMap<String, Value>,
+) -> Result<(), ManyError> {
+    // Make sure we have all the parameters we need for this migration
+    let params = [
+        "symbol",
+        "symbol_name",
+        "symbol_decimals",
+        "symbol_total",       // TODO: This field could be calculated from the storage
```

Comment:
TODO unnecessary. Denormalizing your data in KVStore patterns is not unusual. If you want to keep a comment explaining why, go ahead. KVStore don't have the join query speed of RDBMs.